File: src/fuzz.rs
use std::cmp::Reverse;

use crate::item::{self, Item};

pub struct Fuzz {
    items: Vec<String>,
}

impl Fuzz {
    pub fn new(items: &[Item]) -> Self {
        let mut processed_items = Vec::new();
        for item in items {
            let mut item_text = String::new();
            for part in item.parts {
                item_text.push_str(match part {
                    item::Part::Highlighted { contents } => &contents,
                    item::Part::Plain { contents } => &contents,
                });
            }
            item_text.shrink_to_fit();
            processed_items.push(item_text);
        }
        processed_items.shrink_to_fit();
        Self { items: processed_items }
    }

    pub fn filter(&self, query: &str) -> impl Iterator<Item = item::Index> {
        let mut matches = Vec::new();
        for (index, item_text) in self.items.iter().enumerate() {
            let searcher = sublime_fuzzy::FuzzySearch::new(query, item_text).case_insensitive();
            if let Some(match_) = searcher.best_match() {
                matches.push((index, match_));
            }
        }
        matches.sort_unstable_by_key(|(_index, match_)| Reverse(match_.score()));
        matches.into_iter().map(|(index, _match)| index)
    }
}

File: src/fuzz.py
from typing import Iterable, List
from .item import HighlightedPart, Item, PlainPart
from . import item


class Fuzz:
    def __init__(self, items: List[Item]):
        processed_items = []
        for item in items:
            item_text = ""
            for part in item.parts:
                if isinstance(part, HighlightedPart):
                    item_text += part.content
                elif isinstance(part, PlainPart):
                    item_text += part.content
            processed_items.append(item_text)
        self.items = processed_items

    def filter(self, query: str) -> Iterable[item.Index]:
        matches = []
        for index, item_text in enumerate(self.items):
